File: backend/app/schemas/poi.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, field_validator

from app.models.poi import POIType, POIStatus


class LatLng(BaseModel):
    lat: float
    lng: float


class BBox(BaseModel):
    west: float
    south: float
    east: float
    north: float

    @field_validator("west", "east")
    @classmethod
    def lng_range(cls, v: float) -> float:
        if not -180 <= v <= 180:
            raise ValueError("longitude must be between -180 and 180")
        return v

    @field_validator("south", "north")
    @classmethod
    def lat_range(cls, v: float) -> float:
        if not -90 <= v <= 90:
            raise ValueError("latitude must be between -90 and 90")
        return v


class POIRead(BaseModel):
    id: uuid.UUID
    poi_type: POIType
    location: LatLng
    name: str | None = None
    attributes: dict | None = None
    source: str
    status: POIStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class POIDetail(POIRead):
    """Full detail view including importer metadata (1.3.7)."""

    external_id: str | None = None
    last_verified_at: datetime | None = None
    verification_count: int = 0


class POIListResponse(BaseModel):
    items: list[POIRead]
    truncated: bool

File: frontend/src/types/poi.ts
export type POIType = 'toilet' | 'trash_can' | 'bench' | 'smoking_area' | 'water_fountain'

export interface LatLng {
  lat: number
  lng: number
}

export interface POI {
  id: string
  poi_type: POIType
  location: LatLng
  name: string | null
  attributes: Record<string, unknown> | null
  source: string
  status: string
  created_at: string
  updated_at: string
}

export interface POIDetail extends POI {
  external_id: string | null
  last_verified_at: string | null
  verification_count: number
}

export interface POIListResponse {
  items: POI[]
  truncated: boolean
}

export interface BBox {
  west: number
  south: number
  east: number
  north: number
}

export const POI_TYPE_LABELS: Record<POIType, string> = {
  toilet: '화장실',
  trash_can: '쓰레기통',
  bench: '벤치',
  smoking_area: '흡연구역',
  water_fountain: '음수대',
}

export const POI_TYPE_ICONS: Record<POIType, string> = {
  toilet: '🚻',
  trash_can: '🗑️',
  bench: '🪑',
  smoking_area: '🚬',
  water_fountain: '🚰',
}

export const ALL_POI_TYPES: POIType[] = [
  'toilet', 'trash_can', 'bench', 'smoking_area', 'water_fountain',
]
